agents/self_evolving_car/train.py

```diff
@@ -0,0 +1,163 @@
+
+        # LOOPS
+        self.frame = self.frame + 1
+
+        if self.frame >= self.max_frames:
+            self.frame = 0
+            self.calc_fitness()
+            self.brain += 1  # change bot every reset
+            self.controller_state = SimpleControllerState()  # reset controller
+            self.reset()  # reset at start
+
+        if self.brain >= self.pop:
+            self.gen += 1
+            self.brain = 0  # reset bots after all have gone
+
+            self.avg_best_fitness()
+            self.calc_fittest()
+
+            # PRINT GENERATION INFO
+            print("")
+            print("     GEN = "+str(self.gen))
+            print("-------------------------")
+            print("FITTEST = BOT " + str(self.fittest))
+            print("------FITNESS = " + str(self.bot_fitness[self.fittest]))
+            # print("------WEIGHTS = " + str(self.bot_list[self.fittest]))
+            for i in range(len(self.bot_list)):
+                print("FITNESS OF BOT " + str(i) + " = " + str(self.bot_fitness[i]))
+
+            # NE Functions
+
+            self.selection()
+            self.mutate()
+            self.reset()  # reset because of delay
+
+        return self.controller_state
+
+    def calc_fitness(self):
+        # CALCULATE MINIMUM DISTANCE TO BALL FOR EACH GENOME
+        min_distance_to_ball = min(self.distance_to_ball)
+        self.bot_fitness[self.brain] = min_distance_to_ball
+
+        self.distance_to_ball = [math.inf] * self.max_frames
+
+        return min_distance_to_ball
+
+    def avg_best_fitness(self):
+        # CALCULATE AVG FITNESS OF 5 FITTEST (IDENTICAL) GENOMES
+        self.bot_fitness[-self.num_best:] = [sum(self.bot_fitness[-self.num_best:]) / self.num_best] * self.num_best
+
+    def calc_fittest(self):
+        temp = math.inf
+        for i in range(len(self.bot_list)):
+            if self.bot_fitness[i] < temp:
+                temp = self.bot_fitness[i]
+                self.fittest = i
+        return self.fittest
+
+    def reset(self):
+        # RESET TRAINING ATTRIBUTES AFTER EACH GENOME
+        ball_state = BallState(Physics(velocity=Vector3(0, 0, 0), location=Vector3(self.pos, 5000, 3000),
+                                       angular_velocity=Vector3(0, 0, 0)))
+        car_state = CarState(jumped=False, double_jumped=False, boost_amount=33,
+                             physics=Physics(velocity=Vector3(0,0,0),rotation=Rotator(45, 90, 0),location=Vector3(0.0, -4608,500),angular_velocity=Vector3(0, 0, 0)))
+        game_info_state = GameInfoState(game_speed=1)
+        game_state = GameState(ball=ball_state, cars={self.index: car_state}, game_info=game_info_state)
+        self.set_game_state(game_state)
+
+    def selection(self):
+        # COPY FITTEST WEIGHTS TO ALL GENOMES
+        state_dict = self.bot_list[self.fittest].state_dict()
+        for bot in self.bot_list:
+            bot.load_state_dict(state_dict)
+
+    def mutate(self):
+        # MUTATE FIRST GENOMES
```

Comment:
use python style comments
also what does first genomes mean?
can you explain this better?